pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.query.aggregation.function;
+
+import com.clearspring.analytics.stream.cardinality.CardinalityMergeException;
+import com.clearspring.analytics.stream.cardinality.HyperLogLog;
+import com.linkedin.pinot.common.Utils;
+import com.linkedin.pinot.common.data.FieldSpec.DataType;
+import com.linkedin.pinot.common.request.AggregationInfo;
+import com.linkedin.pinot.core.common.Block;
+import com.linkedin.pinot.core.query.aggregation.AggregationFunction;
+import com.linkedin.pinot.core.query.aggregation.CombineLevel;
+import java.io.Serializable;
+import java.util.List;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Class to implement the old 'fasthll' aggregation function,
+ * only combine and reduce related methods are used,
+ * calling other methods will throw UnsupportedOperationException.
+ *
+ * fasthll takes advantage of pre-aggregated results for fast distinct count estimation
+ *
+ * TODO: fully deprecate and remove this class when new aggregation routine is finished.
+ */
+public class FastHllAggregationFunction implements AggregationFunction<HyperLogLog, Long> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(FastHllAggregationFunction.class);
+
+  private String _distinctCountHLLColumnName;
+
+  @Override
+  public void init(AggregationInfo aggregationInfo) {
+    _distinctCountHLLColumnName = aggregationInfo.getAggregationParams().get("column");
+  }
+
+  @Override
+  public HyperLogLog aggregate(Block docIdSetBlock, Block[] block) {
+    throw new UnsupportedOperationException("Calling old aggregate method of FastHllAggregationFunction is deprecated.");
+  }
+
+  @Override
+  public HyperLogLog aggregate(HyperLogLog mergedResult, int docId, Block[] block) {
+    throw new UnsupportedOperationException("Calling old aggregate method of FastHllAggregationFunction is deprecated.");
+  }
+
+  @Override
+  public List<HyperLogLog> combine(List<HyperLogLog> aggregationResultList, CombineLevel combineLevel) {
+    if ((aggregationResultList == null) || aggregationResultList.isEmpty()) {
+      return null;
+    }
+    HyperLogLog hllResult = aggregationResultList.get(0);
+    for (int i = 1; i < aggregationResultList.size(); ++i) {
+      try {
+        hllResult.addAll(aggregationResultList.get(i));
```

Comment:
This is the old version aggregation function, it just mimic the structure of combine of old distinctcounthll and other functions to have a pre-check of the null value. I'm not sure it's absolutely necessary, but it's always safe to do that.